File: sqrl/sqrl_crypto.py
import hashlib
import hmac
import scrypt
import nacl.bindings
import nacl.signing
import nacl.utils


def _bxor(b1, b2):
    result = bytearray(b1)
    for i, b in enumerate(b2):
        result[i] ^= b
    return bytes(result)

def EnScrypt(password, salt, N, iteration_count):
    """ Working test of test's
    print(binascii.hexlify(EnScrypt('', '', 512, 1)))
    print(binascii.hexlify(EnScrypt('', '', 512, 100)))
    print(binascii.hexlify(EnScrypt('', '', 512, 1000)))
    print(binascii.hexlify(EnScrypt('password', '', 512, 123)))
    print(binascii.hexlify(EnScrypt('password',
        binascii.unhexlify('0000000000000000000000000000000000000000000000000000000000000000'), 512, 123)))

    """

    r = 256
    p = 1 # disable parallelism

    key = None
    for _ in range(iteration_count):
        salt = scrypt.hash(password, salt, N=N, r=r, p=p, buflen=32)
        if key is None:
            key = salt
        else:
            key = _bxor(salt, key)
        #print(binascii.hexlify(key), binascii.hexlify(salt))

    return key

def EnHash(data):
    digest = None
    for _ in range(16):
        data = hashlib.sha256(data).digest()
        if digest is None:
            digest = data
        else:
            digest = _bxor(data, digest)
    return digest

def sqrl_hmac(key, msg):
    if isinstance(msg, str):
        msg = msg.encode()
    return hmac.new(key, msg, hashlib.sha256).digest()

def sqrl_get_ins_from_sin(ssk, sin):
    return sqrl_hmac(EnHash(ssk[:32]), sin)

def sqrl_keypair(raw_sign_key):
    verify, sign = nacl.bindings.crypto_sign_seed_keypair(raw_sign_key)
    # TODO: Is raw_sign_key == sign[:32] always true?
    return verify, sign

def sqrl_make_public(priv):
    return nacl.bindings.crypto_scalarmult_base(priv)

def sqrl_idlock_keys(ilk):
    rlk = nacl.utils.random(size=32)
    print('rlk', rlk)
    suk = sqrl_make_public(rlk)
    dhka = nacl.bindings.crypto_scalarmult(rlk, ilk)
    vuk, ursk = sqrl_keypair(dhka)
    return suk, vuk

def sqrl_get_unlock_request_signing_key(suk, iuk):
    dhka = nacl.bindings.crypto_scalarmult(iuk, suk)
    vuk, ursk = sqrl_keypair(dhka)
    # TODO: Is dhka == ursk always true?
    # TODO: Don't really need to return vuk...
    return ursk, vuk

def sqrl_get_imk_from_iuk(iuk):
    return EnHash(iuk)

def sqrl_get_ilk_from_iuk(iuk):
    return sqrl_make_public(iuk)

def sqrl_get_idk_for_site(imk, sks):
    idk, ssk = sqrl_keypair(sqrl_hmac(imk, sks))
    return idk, ssk

def sqrl_sign(signing_key, message):
    raw_signed = nacl.bindings.crypto_sign(message, signing_key)
    return raw_signed[:nacl.bindings.crypto_sign_BYTES]

File: nrf_sqrl/sqrl_crypto.c
#include "sqrl_crypto.h"

#include <string.h>

#include "app_error.h"
#include "occ_curve25519.h"
#include "occ_ed25519.h"
#include "occ_sha256.h"
#include "occ_hmac_sha256.h"

void sqrl_get_idk_for_site(uint8_t idk[32], uint8_t ssk[32], const uint8_t imk[32], const char* sks)
{
    occ_hmac_sha256(ssk, imk, 32, (uint8_t *)sks, strlen(sks));
    occ_ed25519_public_key(idk, ssk);
}

void sqrl_make_public(uint8_t publ[32], const uint8_t priv[32])
{
    occ_curve25519_scalarmult_base(publ, priv);
}

void sqrl_get_ilk_from_iuk(uint8_t ilk[32], uint8_t iuk[32])
{
    sqrl_make_public(ilk, iuk);
}

void sqrl_get_unlock_request_signing_key(uint8_t ursk[32], uint8_t vuk[32], const uint8_t suk[32], const uint8_t iuk[32])
{
    occ_curve25519_scalarmult(ursk, iuk, suk);
    occ_ed25519_public_key(vuk, ursk);
}

void EnHash(uint8_t digest[32], const uint8_t data[32])
{
    uint8_t tmp_data[32];
    memcpy(tmp_data, data, sizeof(tmp_data));

    memset(digest, 0, 32);
    for (int i = 0; i < 16; i++)
    {
        occ_sha256(tmp_data, tmp_data, sizeof(tmp_data)); // TODO: Safe to reuse tmp_data here?

        for (int j = 0; j < 32; j++) {
            digest[j] ^= tmp_data[j];
        }
    }
}


void sqrl_get_ins_from_sin(uint8_t ins[32], const uint8_t ssk[32], const char* sin)
{
    uint8_t tmp[32];
    EnHash(tmp, ssk);
    occ_hmac_sha256(tmp, ins, 32, (uint8_t *)sin, strlen(sin));
}


void sqrl_idlock_keys(uint8_t suk[32], uint8_t vuk[32], const uint8_t rlk[32], const uint8_t ilk[32])
{
    uint8_t dhka[32];
    sqrl_make_public(suk, rlk);
    occ_curve25519_scalarmult(dhka, rlk, ilk);
    occ_ed25519_public_key(vuk, dhka);
}


void sqrl_crypto_init(void)
{
}
